src/htm/encoders/SimHashDocumentEncoder.cpp

```diff
@@ -0,0 +1,276 @@
+
+/** @file
+ * SimHashDocumentEncoder.cpp
+ * @since 0.2.3
+ */
+
+#include <bitset>
+#include <climits> // CHAR_BIT
+
+#include <hasher.hpp> // digestpp: sha3+shake256 hash digests
+#include <algorithm/sha3.hpp>
+#include <algorithm/shake.hpp>
+
+#include <htm/encoders/SimHashDocumentEncoder.hpp>
+
+
+namespace htm {
+
+  /**
+   * Constructor
+   *
+   * @see SimHashDocumentEncoder.hpp
+   */
+  SimHashDocumentEncoder::SimHashDocumentEncoder(const SimHashDocumentEncoderParameters &parameters)
+  {
+    initialize(parameters);
+  }
+
+  /**
+   * Initialize
+   *
+   * @see SimHashDocumentEncoder.hpp
+   */
+  void SimHashDocumentEncoder::initialize(const SimHashDocumentEncoderParameters &parameters)
+  {
+    // Pre-processing, Check parameters
+    NTA_CHECK((parameters.activeBits > 0u) || (parameters.sparsity > 0.0f))
+      << "Need only one argument of: 'activeBits' or 'sparsity'.";
+    NTA_CHECK(parameters.size > 0u)
+      << "Missing 'size' argument.";
+
+    // Make local copy of arguments for processing
+    args_ = parameters;
+
+    // process: handle 'sparsity' param instead of 'activeBits' param
+    if(args_.sparsity > 0.0f) {
+      NTA_CHECK((args_.sparsity >= 0.0f) && (args_.sparsity <= 1.0f))
+        << "Argument 'sparsity' must be a float in the range 0.0-1.0.";
+      NTA_CHECK(args_.size > 0u)
+        << "Argument 'sparsity' requires that the 'size' also be given.";
+      args_.activeBits = (UInt) round(args_.size * args_.sparsity);
+    }
+
+    // Post-processing, Sanity check the parameters
+    NTA_CHECK(args_.size > 0u);
+    NTA_CHECK(args_.activeBits > 0u);
+    NTA_CHECK(args_.activeBits < args_.size);
+
+    // Initialize parent class with finalized params
+    BaseEncoder<std::vector<std::string>>::initialize({ args_.size });
+  } // end method initialize
+
+  /**
+   * Encode (Main calling style)
+   *
+   * If param "tokenSimilarity" is passed, we'll additionally loop through
+   *  all the letters in all the tokens.
+   * Each token/letter will be hashed with SHA3+SHAKE256 to get
+   *  a varible-length (param "size") binary digest output. These vectors will
+   *  be stored in a matrix for the next step of processing.
+   * Weights are added in during hashing and simhashing.
+   * After the loop, we SimHash the matrix of hashes, resulting in an
+   *  output SDR.
+   *
+   * @param :input: Document token strings (with weights) to encode,
+   *  ex: {{ "what", 3 }, { "is", 1 }, { "up", 2 }}.
+   * @param :output: Result SDR to fill with result output encoding.
+   *
+   * @see SimHashDocumentEncoder.hpp
+   * @see encode(const std::vector<std::string> input, SDR &output)
+   */
+  void SimHashDocumentEncoder::encode(const std::map<std::string, UInt> input, SDR &output)
+  {
+    NTA_CHECK(input.size() > 0u)
+      << "Encoding input vector array should have at least 1 string member.";
+
+    Eigen::MatrixXi adders(args_.size, 0u);
```

Comment:
nit, using Eigen is fine. I have included the lib in the repo. On the other hand, currently no other code depends on eigen, so if this does not offer a significant performance gain and can be done using std c++ (vector, map) than that would be probably better.

Reply:
> Nothing I couldn't re-write by hand, but it sure is nicer this way. :)

fine to leave as is now with eigen, then.